fix stuck floordiv

avikchaudhuri opened the pull request:
Summary: Fixes https://github.com/pytorch/pytorch/issues/134133

Test Plan:
Tested on the small repro in the linked issue with different lengths N (replacing 100), recording N vs. time taken in nanoseconds:
10 127268319
20 220839662
30 325463125
40 429259441
50 553136055
60 670799769
70 999170514
80 899014103
90 997168902
100 1168202035
110 1388556619
120 1457488235
130 1609816470
140 2177889877
150 1917560313
160 2121096113
170 2428502334
180 4117450755
190 4003068224

So N ~ 200 takes ~5s. Previously even smaller N would go for >1 min.

Didn't add a perf test because ezyang is planning to build a benchmark.

Also tested on https://www.internalfb.com/diff/D61560171, which now gets past the stuck point.

Differential Revision: D61619660


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

isuruf commented on torch/utils/_sympy/functions.py:
```diff
@@ -156,8 +157,9 @@ def eval(cls, base, divisor):
-            gcd = sympy.gcd(base, divisor)
-            if not equal_valued(gcd, 1):
+            # sympy.gcd tends to blow up on large sums, so use it on each summand instead
+            gcd, *gcds_ = [sympy.gcd(term, divisor) for term in base_args]
+            if not equal_valued(gcd, 1) and all(equal_valued(gcd, gcd_) for gcd_ in gcds_):
```
These two are not equivalent right?

avikchaudhuri replied:
Yeah I don't think they are equivalent in general, e.g. `gcd(x + 1, x + 1) = x + 1` but `gcd(x, x + 1)` and `gcd(1, x + 1)` are both `1`. Another example: `gcd(2x + 3x, 5)`. 

I actually don't know how far `sympy.gcd` goes, but the net effect here is that we'll simplify fewer expressions that could have possibly been simplified previously.

ezyang replied:
It's a little annoying that if I have something like `(2 * x + 4 * y) // 8` I won't manage to simplify it anymore but I guess if the tests pass tis fine

avikchaudhuri replied:
Indeed annoying. Hard to tell without data on what kind of simplifications are actually enabled in practice by `sympy.gcd`, but if its just integer factors being cancelled out then we could do something simple like: https://github.com/pytorch/pytorch/blob/7af38eb98bdceb8fc6f8635ed7dd664ef44e4b10/torch/fx/experimental/symbolic_shapes.py#L427

ezyang reviewed (approved):
Can you also add a benchmark for this in the new microbenchmark section? `benchmarks/dynamo/pr_time_benchmarks/benchmarks/update_hint_benchmark.py` as an example. cc @laithsakka

avikchaudhuri committed:
fix stuck floordiv (#134150)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134150

Fixes https://github.com/pytorch/pytorch/issues/134133

Test Plan:
Tested on the small repro in the linked issue with different lengths N (replacing 100), recording N vs. time taken in nanoseconds:
10 127268319
20 220839662
30 325463125
40 429259441
50 553136055
60 670799769
70 999170514
80 899014103
90 997168902
100 1168202035
110 1388556619
120 1457488235
130 1609816470
140 2177889877
150 1917560313
160 2121096113
170 2428502334
180 4117450755
190 4003068224

So N ~ 200 takes ~5s. Previously even smaller N would go for >1 min.

Didn't add a perf test because ezyang is planning to build a benchmark.

Also tested on https://www.internalfb.com/diff/D61560171, which now gets past the stuck point.

Reviewed By: ezyang

Differential Revision: D61619660

ezyang commented:
Is there more information about this somewhere? There is no comment on the diff

avikchaudhuri commented:
> Is there more information about this somewhere? There is no comment on the diff

https://fb.workplace.com/groups/1075192433118967/posts/1491493248155548?comment_id=1491938994777640

avikchaudhuri commented:
@ezyang I'm trying to run the internal test with logging to see what we might have computed differently and I see, e.g., :
```
W0829 102853.039 functions.py:206] COMPUTING GCD 8192*((xindex//256)) + 128*ModularIndexing(xindex, 4, 64) // 65536
W0829 102853.041 functions.py:208] COMPUTED GCD 128 of 8192*((xindex//256)) + 128*ModularIndexing(xindex, 4, 64) // 65536
W0829 102853.041 functions.py:210] GCD 128 != 1
W0829 102853.054 functions.py:213] SIMPLIFIED 64*((xindex//256)) + ModularIndexing(xindex, 4, 64) // 512
```

So it comes down to finding common integer factors and I have some code to do that surgically, I'll send another PR after some more testing...

ezyang commented:
At least we did catch this in code review :P

avikchaudhuri commented:
Alternate PR.
